freedb/protocols.md

```diff
@@ -1,8 +1,311 @@
+]
+```
+
+Then the actual data that we store in Google Sheet would look like this:
+| _rid | name | age |
+| ---- | ---- | --- |
+| 1    | A    | 1   |
+| 2    | B    | 2   |
+| 3    | C    | 3   |
+| 4    | D    | 4   |
+
+#### About `_rid` Metadata
+
+`_rid` column will be used to tell the index of a row. This metadata will make our Google Sheet Formula simpler
+so that we can utilize [Google Visualization API][GVizAPI] for our needs.
+
+⚠️ Even though `_rid` can be used as the row identifier (because `_rid` is unique between each row), we recommend the
+client to implement their own row identifier mechanism such as [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+or [Snowflake](https://en.wikipedia.org/wiki/Snowflake_ID) because we will not guarantee that `_rid` will not be
+reused out-of-the-box (to ensure the `_rid` is not reused, you need to implement soft-delete mechanism by your own).
+
+Consider this scenario:
+1. Client inserts Row A to the store, which will have `_rid=2`.
+2. Client deletes Row A from the store.
+3. Client Insert Row B to the store, which will reuse the same `_rid` as A i.e. B's `_rid` will be equal to `2`.
+
+We did not expose this field by default because if the client didn't aware of this behavior it can cause various bugs
+because of the invalid/dangling IDs.
+
+### Operations
+
+#### Insert
+
+Inserts new rows into the database.  The operation method should expect these parameters:
+| Parameter Name | Required | Remarks                                                   |
+| -------------- | -------- | --------------------------------------------------------- |
+| `rows`         | Yes      | List of row objects that will be inserted into the store. |
+
+
+Call [spreadsheets.values.append][AppendAPI]
+API with the following parameters:
+
+| Parameter Name              | Value                      |
+| --------------------------- | -------------------------- |
+| `spreadsheetId`             | `<current_spreadsheet_id>` |
+| `range`                     | `<data_range>`             |
+| `responseValueRenderOption` | `FORMATTED_VALUE`          |
+| `valueInputOption`          | `client_ENTERED`           |
+| `includeValuesInResponse`   | `true`                     |
+| `insertDataOption`          | `OVERWRITE`                |
+
+Replace the `<current_spreadsheet_id>` with the Google Spreadsheet ID that we're going to operate on and `<data_range>`
+with the range where the data lives in [A1 notation format][A1Notation]. It's recommended to select the entire column
+for the `<data_range>`, e.g. suppose we have 5 columns (including `_rid` metadata column) in `Sheet1` we should
+use `Sheet1!A:E` to select the entire rows.
+
+To construct the value of `values` parameter (inside the request's body), convert each row into an array where the
+1st value (value of the `_rid` metadata column) equals to `"=ROW()"` and `i`-th (1-based index) value contains the
+value of `i`-th column of the current row.
```

Comment:
This statement seems confusing for the `i-th` logic. I think we should use both 1-based index for both the value in the `values` row array vs the column from the struct/class.